### DISCUSSION TITLE
Runtime error while trying to turnoff advanced minibatching

### DISCUSSION THREAD

**Discussion by Zevrap-81** [pyg-team/pytorch_geometric#5795] [Q&A]
I am getting the following error when i set the function __cat_dim__ to return None
'''
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch\utils\data\dataloader.py", line 681, in __next__
    data = self._next_data()
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch\utils\data\dataloader.py", line 721, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch\utils\data\_utils\fetch.py", line 52, in fetch
    return self.collate_fn(data)
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch_geometric\loader\dataloader.py", line 19, in __call__
    return Batch.from_data_list(batch, self.follow_batch,
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch_geometric\data\batch.py", line 76, in from_data_list
    batch, slice_dict, inc_dict = collate(
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch_geometric\data\collate.py", line 107, in collate
    repeats = [store.num_nodes for store in stores]
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch_geometric\data\collate.py", line 107, in <listcomp>
    repeats = [store.num_nodes for store in stores]
  File "D:\Users\boos\Parvez\Neuer Ordner\graph-sage\env\lib\site-packages\torch_geometric\data\storage.py", line 266, in num_nodes
    return value.size(self._parent().__cat_dim__(key, value, self))
RuntimeError: Please look up dimensions by name, got: name = None.
'''
A minimal code to reproduce the error is given here:
```
import os.path as osp
import torch

from torch_geometric.data import Data, Dataset
from torch_geometric.loader import DataLoader


class _Data(Data):
    def __cat_dim__(self, key, value, *args, **kwargs):
        if key == 'x' or key == 'y':
            return None
        else:
            return super().__cat_dim__(key, value, *args, **kwargs)

class MyDataset(Dataset):
    def __init__(self):
        self.num_samples= 10
        self.data_dir= 'rough_data'
        super().__init__(root= self.data_dir)

    def process(self):
        for i in range(self.len()):
            x= torch.rand(10,5)
            y= torch.rand(10,1)
            data= _Data(x=x,y=y)
            torch.save(data, osp.join(self.processed_dir, rf"data_{i}.pt"))
    
    def len(self):
        return self.num_samples
    
    def get(self, idx):
        data= torch.load(osp.join(self.processed_dir, rf"data_{idx}.pt"))
        return data 

    @property
    def processed_dir(self) -> str:
        return osp.join(self.data_dir, r'Data_after_preprocessing')   
       
    @property
    def processed_file_names(self):
        r"""The name of the files in the :obj:`self.processed_dir` folder that
        must be present in order to skip processing."""
        return [r"data_0.pt", rf"data_{self.len()-1}.pt"]

if __name__ == "__main__":
    dataset= MyDataset()
    loader= DataLoader(dataset, batch_size=5)
    for data in loader:
        print(loader.x.shape)
        break
```

**Comment by EdisonLeeeee** [comment]
Any reason to set `cat_dim ` to `None`? IMO `x` and `y` should be contacted along the first dimension (i.e., `cat_dim=0`).

**Reply by Zevrap-81** [reply]
I want them to concatenate along a new dimension

**Reply by EdisonLeeeee** [reply]
Can you clarify? If I understood it right, you might need the `torch.stack` instead of `torch.cat` to concatenate them, but it is not supported in PyG DataLoader.

**Reply by Zevrap-81** [reply]
But in this [pull request](https://github.com/pyg-team/pytorch_geometric/pull/2036) it shows that it is possible:

**Reply by EdisonLeeeee** [reply]
Didn't notice that, my bad. That seems to be a bug. Will investigate it further and report back.

**Reply by rusty1s** [reply]
That is indeed tricky. One easy way to fix this is to just manually specify `data.num_nodes`. Would that work for you?

**Reply by Zevrap-81** [reply]
yes this works for me. But i wanted to ask if this classical batching will cause any unwanted problems with the models from torch_geometric. I mean if my x is of shape [bs, num_nodes, channels ] but my edge_index is of shape [2, bs*num_edges]?

**Reply by rusty1s** [reply]
There exists a handful of operator that are tested to work on this (`GCNConv`, `GraphConv`, `SAGEConv`, …), but it may not work for all of them (e.g., `GATConv`).
